File: base/base-client.ts
import { z } from "zod";

function createApiResponse<
  Data extends z.ZodTypeAny,
  Errors extends z.ZodTypeAny,
>(data: Data, errors: Errors) {
  const error = z.object({
    code: z.number(),
    message: z.string(),
    errors,
  });

  return z.discriminatedUnion("status", [
    z.object({ status: z.literal("success"), data }),
    z.object({ status: z.literal("error"), error }),
  ]);
}

export type ExtraOptions = {
  headers?: Record<string, string>;
  query?: Record<string, string>;
};

export class BaseApiClient {
  baseUrl: string;
  token?: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  setToken(token?: string) {
    this.token = token;
  }

  async request<T extends z.ZodTypeAny>(
    endpoint: string,
    method: string,
    bodySchema: T,
    body?: any,
    extra?: ExtraOptions,
  ) {
    const headers: Record<string, string> = {};
    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }

    if (body) {
      headers["Content-Type"] = "application/json";
    }

    const url = new URL(this.baseUrl + endpoint);

    if (extra) {
      if (extra.headers) {
        for (const [key, value] of Object.entries(extra.headers)) {
          headers[key] = value;
        }
      }

      if (extra.query) {
        for (const [key, value] of Object.entries(extra.query)) {
          url.searchParams.set(key, value);
        }
      }
    }

    const res = await fetch(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : null,
    });

    const Schema = createApiResponse(bodySchema, z.undefined());

    const data = await res.json();
    const parsedData = await Schema.parseAsync(data);

    return parsedData;
  }
}

File: spark/dart/base_client.dart
import 'dart:convert';

import 'package:result_dart/result_dart.dart';
import 'package:dio/dio.dart';

class ApiError {
  const ApiError(this.type, this.code, this.message);

  final String type;
  final int code;
  final String message;
}

class NoBody {}

class RequestOptions {
  Map<String, dynamic>? headers;
  Map<String, dynamic>? query;

  RequestOptions({this.query, this.headers});
}

class BaseApiClient {
  BaseApiClient({String baseUrl = ""}) {
    _dio = Dio(
      BaseOptions(
        baseUrl: baseUrl,
        validateStatus: (status) {
          return true;
        },
      ),
    );

    headers = <String, String>{};
  }

  late Dio _dio;
  late Map<String, String> headers;

  AsyncResultDart<Map<String, dynamic>, ApiError> request(
    String method,
    String path, {
    RequestOptions? options,
    Map<String, dynamic>? body,
  }) async {
    final headers = <String, dynamic>{...this.headers};
    headers["Content-Type"] = "application/json";

    if (options?.headers != null) {
      headers.addAll(options!.headers!);
    }

    final res = await _dio.request(
      path,
      options: Options(method: method, headers: headers),
      queryParameters: options?.query,
      data: body != null ? jsonEncode(body) : null,
    );

    final data = res.data as Map<String, dynamic>;

    final success = data["success"] as bool;

    if (!success) {
      final error = data["error"] as Map<String, dynamic>;
      return Failure(
        ApiError(
          error["type"] as String,
          error["code"] as int,
          error["message"] as String,
        ),
      );
    }

    return Success(data["data"]);
  }
}
